zebra-network/src/peer/connection.rs

```diff
@@ -449,6 +527,78 @@ where
+                    .iter()
+                    .all(|item| matches!(item, InventoryHash::Block(_))) =>
+            {
+                Some(Request::BlocksByHash(
+                    items
+                        .iter()
+                        .map(|item| {
+                            if let InventoryHash::Block(hash) = item {
+                                *hash
+                            } else {
+                                unreachable!("already checked all items are InventoryHash::Block")
+                            }
+                        })
+                        .collect(),
+                ))
+            }
+            Message::GetData(items)
+                if items
+                    .iter()
+                    .all(|item| matches!(item, InventoryHash::Tx(_))) =>
+            {
+                Some(Request::TransactionsByHash(
+                    items
+                        .iter()
+                        .map(|item| {
+                            if let InventoryHash::Tx(hash) = item {
+                                *hash
+                            } else {
+                                unreachable!("already checked all items are InventoryHash::Tx")
+                            }
+                        })
+                        .collect(),
+                ))
+            }
+            Message::GetData(items) => {
+                debug!(?items, "could not interpret getdata message");
+                None
+            }
+            Message::Tx(transaction) => Some(Request::PushTransaction(transaction)),
+            // We don't expect to be advertised multiple blocks at a time,
+            // so we ignore any advertisements of multiple blocks.
+            Message::Inv(items)
+                if items.len() == 1 && matches!(items[0], InventoryHash::Block(_)) =>
+            {
+                if let InventoryHash::Block(hash) = items[0] {
+                    Some(Request::AdvertiseBlock(hash))
+                } else {
+                    unreachable!("already checked we got a single block hash");
+                }
+            }
+            // This match arm is terrible, because we have to check that all the items
+            // are the correct kind and *then* convert them all.
+            Message::Inv(items)
+                if items
+                    .iter()
+                    .all(|item| matches!(item, InventoryHash::Tx(_))) =>
+            {
+                Some(Request::AdvertiseTransactions(
+                    items
+                        .iter()
+                        .map(|item| {
+                            if let InventoryHash::Tx(hash) = item {
+                                *hash
+                            } else {
+                                unreachable!("already checked all items are InventoryHash::Tx")
+                            }
+                        })
+                        .collect(),
+                ))
+            }
```

Comment:
you could collapse this by using `collect` and some error handling internally, this changes the behaviour a bit, so it no longer prints "unhandled message type" if any of them would map to none, you could probably add a `.map_err` before the `ok` to print a similar message if it's important.

```suggestion
            Message::Inv(items) => items
                .iter()
                .map(|item| {
                    if let InventoryHash::Tx(hash) = item {
                        Ok(*hash)
                    } else {
                        Err(())
                    }
                })
                .collect::<Result<_, ()>>()
                .map(Request::AdvertiseTransactions)
                .ok(),
```

Reply:
I think it's important that the match arm matches *only* messages where all of the inventory items are transaction hashes, because otherwise we're moving match logic out of the match statement and into the arms.  The code is superficially uglier, yes, but it's more cleanly factored, because all of the matching criteria are done upfront and the match arm is independent of other match arms.

Reply:
makes sense